phoenix-core-client/src/main/java/org/apache/phoenix/expression/util/bson/UpdateExpressionUtils.java

```diff
@@ -0,0 +1,790 @@
+    }
+  }
+
+  /**
+   * Update the given document by performing DELETE operation. This operation is applicable
+   * only on Set data structure. The document is updated by removing the given set of elements from
+   * the given set of elements.
+   *
+   * @param deleteExpr Delete Expression Document with key-value pairs. Key represents field in the
+   * given document, on which operation is to be performed. Value represents set of elements to be
+   * removed from the existing set.
+   * @param bsonDocument Document contents to be updated.
+   */
+  private static void executeDeleteExpression(final BsonDocument deleteExpr,
+      final BsonDocument bsonDocument) {
+    for (Map.Entry<String, BsonValue> deleteEntry : deleteExpr.entrySet()) {
+      String fieldKey = deleteEntry.getKey();
+      BsonValue newVal = deleteEntry.getValue();
+      BsonValue topLevelValue = bsonDocument.get(fieldKey);
+      if (!isBsonSet(newVal)) {
+        throw new RuntimeException("Type of new value to be removed should be sets only");
+      }
+      if (topLevelValue != null) {
+        BsonValue value = modifyFieldValueByDelete(topLevelValue, newVal);
+        if (value == null) {
+          bsonDocument.remove(fieldKey);
+        } else {
+          bsonDocument.put(fieldKey, value);
+        }
+      } else if (!fieldKey.contains(".") && !fieldKey.contains("[")) {
+        LOGGER.info("Nothing to be removed as field with key {} does not exist", fieldKey);
+      } else {
+        updateNestedFieldEntryByDelete(fieldKey, bsonDocument, newVal);
+      }
+    }
+  }
+
+  private static void updateNestedFieldEntryByDelete(final String fieldKey,
+                                                     final BsonDocument bsonDocument,
+                                                     final BsonValue newVal) {
+    if (fieldKey.contains(".") || fieldKey.contains("[")) {
+      StringBuilder sb = new StringBuilder();
+      for (int i = 0; i < fieldKey.length(); i++) {
+        if (fieldKey.charAt(i) == '.') {
+          BsonValue value = bsonDocument.get(sb.toString());
+          if (value == null) {
+            LOGGER.error("Incorrect access. Should have found nested bsonDocument for {}", sb);
+            throw new RuntimeException("Map does not contain key: " + sb);
+          }
+          updateNestedFieldByDelete(value, i, fieldKey, newVal);
+          return;
+        } else if (fieldKey.charAt(i) == '[') {
+          BsonValue value = bsonDocument.get(sb.toString());
+          if (value == null) {
+            LOGGER.error("Incorrect access. Should have found nested list for {}", sb);
+            throw new RuntimeException("Map does not contain key: " + sb);
+          }
+          updateNestedFieldByDelete(value, i, fieldKey, newVal);
+          return;
+        } else {
+          sb.append(fieldKey.charAt(i));
+        }
+      }
+    }
+  }
+
+  private static void updateNestedFieldByDelete(final BsonValue value, final int idx,
```

Comment:
It will be good to explain the meaning of the parameters and the basic logic in the method.

Reply:
Done